### DISCUSSION TITLE
ProcessToRunonBan - Pass arguments when null

### DISCUSSION THREAD

**Discussion by landon-lengyel** [DigitalRuby/IPBan#285] [Ideas]
When utilizing `ProcessToRunOnBan` with a PowerShell script (and I assume other stuff), the script is unable to run if there is a missing variable. For example,

Consider the config:
```xml
<add key="ProcessToRunOnBan" value="C:\Program Files\PowerShell\7\pwsh.exe|-Command &quot;&amp; 'C:\Program Files\IPBan\Send-OpenSearch.ps1' -Banned 1 -IPAddress ###IPADDRESS### -Username ###USERNAME### -Count ###COUNT### -Source ###SOURCE### -Log ###LOG###&quot;"/>
```

If a scenario in which ###USERNAME### is null, the script will not run spitting out 'Missing an argument for parameter' (though this is not logged by IPBan or anything else)

If the IPBan could instead spit out the string `null` or some other pre-defined string, that would let me write the PowerShell script to handle that scenario. Of course, that should probably be an additional config option as to not interfere with existing deployments.

The only workaround I can think of without this option, would be using one massive parameter and adding some arbitrary delimiter. I'm open to suggestions if someone has a better way though.

**Comment by jjxtra** [comment]
Can powershell be set to allow optional parameters?

**Comment by landon-lengyel** [comment]
It can. The parameters themselves are optional, but if the parameter name is included then it requires some value. 
So I could simply not include a `-Username` option, but then it would never be used and kind of defeats the point. 

I'm probably going to workaround this with the 'big old monster parameter' for now, but it may be something to consider for your next release :)

I appreciate the work you do!

**Comment by jjxtra** [comment]
If the software passed "" around each parameter, would that be sufficient?

**Comment by landon-lengyel** [comment]
I just tested that and it appears that PowerShell will accept empty strings there. So that would resolve this for PowerShell at least!

**Reply by jjxtra** [reply]
I can do a quick update for that

**Reply by landon-lengyel** [reply]
Oh, that's perfect. Thank you! I can test it once you're ready.

**Reply by jjxtra** [reply]
@landon-l8 Actually you can add the quotes to your config file around each ###...### entry. If I add them it will mess people up who may have already added them.

`<add key="ProcessToRunOnBan" value="C:\Program Files\PowerShell\7\pwsh.exe|-Command &quot;&amp; 'C:\Program Files\IPBan\Send-OpenSearch.ps1' -Banned 1 -IPAddress '###IPADDRESS###' -Username '###USERNAME###' -Count '###COUNT###' -Source '###SOURCE###' -Log '###LOG###'&quot;"/>`

**Reply by landon-lengyel** [reply]
I didn't think to try that for some reason, but that works great! Thank you for your help.

**Comment by landon-lengyel** [comment]
Solution is to add single quotes around the IPBan variables, so that they always return as a string (even an empty string). Then PowerShell will always run.

For example:
```xml
<add key="ProcessToRunOnBan" value="C:\Program Files\PowerShell\7\pwsh.exe|-Command &quot;&amp; 'C:\Program Files\IPBan\Send-OpenSearch.ps1' -Banned 1 -IPAddress '###IPADDRESS###' -Username '###USERNAME###' -Count '###COUNT###' -Source '###SOURCE###' -Log '###LOG###'&quot;"/>
```
